File: backend/src/test/java/com/team05/petmeeting/domain/ads/controller/AdsControllerTest.java
package com.team05.petmeeting.domain.ads.controller;

import com.team05.petmeeting.domain.user.entity.User;
import com.team05.petmeeting.domain.user.repository.UserRepository;
import com.team05.petmeeting.global.security.userdetails.CustomUserDetails;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication;

import java.util.List;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc  // 이게 있는지 확인
@Transactional
@ActiveProfiles("test")
public class AdsControllerTest {

    @Autowired
    private UserRepository userRepository;
    private Long userId;

    @Autowired
    private MockMvc mvc;

    @BeforeEach
    void setUp() {
        User user = User.create("test@test.com", "테스터", "홍길동");
        user = userRepository.save(user);
        userId = user.getId();
    }

    private UsernamePasswordAuthenticationToken auth() {
        CustomUserDetails userDetails = new CustomUserDetails(userId, List.of());
        return new UsernamePasswordAuthenticationToken(userDetails, null, List.of());
    }

    @Test
    @DisplayName("Top N 동물 조회 성공")
    void getTopAnimals_success() throws Exception {
        mvc.perform(get("/api/v1/ads/top-animals")
                        .param("n", "3")
                        .with(authentication(auth())))  // 인증 추가
                .andExpect(status().isOk());
    }
}

File: backend/src/test/java/com/team05/petmeeting/domain/ads/controller/AdsControllerTest.kt
package com.team05.petmeeting.domain.ads.controller

import com.team05.petmeeting.domain.ads.service.AdsService
import com.team05.petmeeting.global.security.handler.JwtAuthenticationEntryPoint
import com.team05.petmeeting.global.security.util.JwtUtil
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.handler
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@WebMvcTest(AdsController::class)
@AutoConfigureMockMvc(addFilters = false)
internal class AdsControllerTest {
    @Autowired
    private lateinit var mvc: MockMvc

    @MockitoBean
    private lateinit var adsService: AdsService

    @MockitoBean
    private lateinit var jwtUtil: JwtUtil

    @MockitoBean
    private lateinit var jwtAuthenticationEntryPoint: JwtAuthenticationEntryPoint

    @Test
    @DisplayName("Top N 동물 조회 성공")
    fun getTopAnimals_success() {
        Mockito.`when`(adsService.getTopAnimals(3)).thenReturn(emptyList())

        mvc.perform(
            get("/api/v1/ads/top-animals")
                .param("n", "3")
        )
            .andExpect(handler().handlerType(AdsController::class.java))
            .andExpect(handler().methodName("getTopAnimals"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.msg").value("Top N 동물 조회 성공"))
            .andExpect(jsonPath("$.resultCode").value("200"))
            .andExpect(jsonPath("$.data").isArray)
    }
}
